### PR TITLE
[MLIR] Bump MLIR commit to c61db4bb

### PR DESCRIPTION
**Pull request by nmostafa**
Fixes:
* Rename Affine loop builders. 
* Parser/Printer call-backs take an AsmParser/Printer. 
* StaticFloatMemRef is removed. Added our own struct. The struct just holds the first data pointer and skips offset, shape and strides. This is OK since we only support static shapes/strides now. 
* Added more unit-tests to cover NGTensorType parsing.

### CODE REVIEW

File: src/contrib/mlir/core/ngraph_dialect/dialect.cpp [hunk 1]
```diff
@@ -39,63 +39,60 @@ NGraphOpsDialect::NGraphOpsDialect(mlir::MLIRContext* ctx)
...
-        {
-            return (emitError(loc, "Unexpected element type in tensor type: " + tyData), Type());
-        }
+        // Parse the current element type.
+        Type eltType;
+
+        parser.parseType(eltType);
 
-        return NGTensorType::get(context, elem_ty, shape);
+        parser.parseGreater();
```
Comment (dcaballe) [inline src/contrib/mlir/core/ngraph_dialect/dialect.cpp]: Nice! These are great improvements! Should we keep error messages in lines 81 and 70?

File: src/contrib/mlir/core/ngraph_dialect/dialect.cpp [hunk 2]
```diff
@@ -39,63 +39,60 @@ NGraphOpsDialect::NGraphOpsDialect(mlir::MLIRContext* ctx)
...
+        return NGTensorType::get(context, eltType, shape);
     }
+    else
+    {
+        // parse nGraph scalar type
+        return parseEltType(parser);
```
Comment (dcaballe) [inline src/contrib/mlir/core/ngraph_dialect/dialect.cpp]: 64 goes through `parser` and 72 calls `parseEltType` directly. Shouldn't we use the same for both?

File: src/contrib/mlir/runtime/cpu/cpu_runtime.cpp [hunk 3]
```diff
@@ -129,39 +129,37 @@ void MLIRCPURuntime::cleanup()
     // Free void double pointer arguments without freeing external tensor data.
     for (auto* arg : m_invokeArgs)
     {
-        auto* memRefArg = *(reinterpret_cast<mlir::StaticFloatMemRef**>(arg));
+        auto* memRefArg = *(reinterpret_cast<StaticMemRef**>(arg));
```
Comment (dcaballe) [inline src/contrib/mlir/runtime/cpu/cpu_runtime.cpp]: How is MLIR replacing `StaticFloatMemRef`?

File: test/mlir/ngraph_dialect/types.mlir [hunk 4]
```diff
@@ -55,23 +55,23 @@ func @i64(%arg0: !ng.i64) {
...
+// CHECK-SAME: (%{{.*}}: !ng.u16)
 func @u16(%arg0: !ng.u16) {
   "ng.return"() : () -> ()
 }
 
 // -----
 
 // CHECK-LABEL: func @u32
-// CHECK-SAME: (%{{.*}}: !ng.i32)
+// CHECK-SAME: (%{{.*}}: !ng.u32)
```
Comment (dcaballe) [inline test/mlir/ngraph_dialect/types.mlir]: Thanks for fixing this

File: src/contrib/mlir/core/ngraph_dialect/dialect.cpp [hunk 5]
```diff
@@ -39,63 +39,60 @@ NGraphOpsDialect::NGraphOpsDialect(mlir::MLIRContext* ctx)
...
+    // failure is true
+    if (!parser.parseOptionalKeyword("tensor"))
     {
-        if (!tyData.consume_front("<") || !tyData.consume_back(">"))
+        llvm::SMLoc typeLoc = parser.getCurrentLocation();
+        if (parser.parseLess())
         {
-            return (emitError(loc, "expected '<' and '>' enclosing the tensor shape: " + tyData),
-                    Type());
+            parser.emitError(typeLoc, "expected '<' and '>' enclosing the tensor shape");
```
[the remark this answers is not in the document]
Comment (nmostafa) [reply src/contrib/mlir/core/ngraph_dialect/dialect.cpp]: No need to. The parser emitError will print the offending line along with a location.

File: src/contrib/mlir/core/ngraph_dialect/dialect.cpp (hunk 2 above)
Comment (nmostafa) [reply src/contrib/mlir/core/ngraph_dialect/dialect.cpp]: The difference is subtle: @line 64 the parser is at `ng.` if we have an nGraph type (e.g ng.i32) or at the type token directly if we have std type (e.g f32). The call to `parser.parseType` will decide whether to call us back or not based on the existence of that prefix. IOW, it is capable of handling both ng and std types. 
@line 72, the parser is already past the `ng.` prefix (if any), when we go the callback. If I call `parser.parseType` here on `ng.i32`, the parser will assume intType token (i32) and return std type. I gave that a try.

File: src/contrib/mlir/runtime/cpu/cpu_runtime.cpp (hunk 3 above)
Comment (nmostafa) [reply src/contrib/mlir/runtime/cpu/cpu_runtime.cpp]: I haven't seen any replacement. JITRunner has its own struct as well. If you know of any, please point me to it.

File: src/contrib/mlir/core/ngraph_dialect/dialect.cpp (hunk 1 above)
Comment (nmostafa) [reply src/contrib/mlir/core/ngraph_dialect/dialect.cpp]: Right. Fixed.

**Pull request merged**